### PR TITLE
SALTO-2674, SALTO-3301, SALTO-2096, SALTO-3439: add fetch warning instead of failing on 403

### PR DESCRIPTION
**Pull request by edenhassid**
_add fetch warning instead of failing on 403_

---

_Additional context for reviewer_

---
_Release Notes_: 
add fetch warning instead of failing on 403

---
_User Notifications_:

### CODE REVIEW

File: packages/adapter-components/src/elements/element_getter.ts [hunk 1]
```diff
@@ -52,8 +54,8 @@ export const getElementsWithContext = async <E extends Element>({
   typeElementGetter: (args: {
     typeName: string
     contextElements?: Record<string, E[]>
-  }) => Promise<E[]>
-}): Promise<E[]> => {
+  }) => Promise<FetchElements<E[]>>
```
Comment (netama) [inline packages/adapter-components/src/elements/element_getter.ts:57]: I think we can just define the `FetchElements` type in this file right before the function - we import from this file both in `ducktype` and in `swagger` in the relevant files, and in general it's better if we don't import anything swagger- or ducktype-specific in the shared code so that it doesn't rely on it.
also - I see we also already have a `FetchElements` type defined under ducktype/transformer.ts - wdyt about moving it here and reusing it for all of these (we can make `configChanges` optional as well, and aggregate in the relevant places with `getUniqueConfigSuggestions`)? if not let's just rename one of them so there's no confusion

File: packages/adapter-components/src/elements/element_getter.ts [hunk 2]
```diff
@@ -75,30 +77,43 @@ export const getElementsWithContext = async <E extends Element>({
 
   const contextElements: Record<string, {
```
Comment (netama) [inline packages/adapter-components/src/elements/element_getter.ts:78]: I think this can be a `Record<string, FetchElements>`

File: packages/adapter-components/src/elements/element_getter.ts [hunk 3]
```diff
@@ -75,30 +77,43 @@ export const getElementsWithContext = async <E extends Element>({
...
+      ]
+    })
   ))
   const dependentElements = await Promise.all(
-    [...dependentEndpoints].map(async typeName => typeElementGetter({
-      typeName,
-      contextElements: _.mapValues(contextElements, val => val.elements),
-    }))
+    [...dependentEndpoints].map(async typeName => {
+      const res = await typeElementGetter({
```
Comment (netama) [inline packages/adapter-components/src/elements/element_getter.ts:98]: why was this change needed, is there something else in `res` that we don't want to return?

File: packages/adapter-components/src/elements/element_getter.ts [hunk 4]
```diff
@@ -75,30 +77,43 @@ export const getElementsWithContext = async <E extends Element>({
...
 
-  return [
-    ...Object.values(contextElements)
-      .flatMap(({ persistInstances, elements }) => (persistInstances
-        ? elements
-        : [])),
-    ...dependentElements.flat(),
-  ]
+  return {
+    elements: [
```
Comment (netama) [inline packages/adapter-components/src/elements/element_getter.ts:107]: not related to your change, but I think we may end up reaching here a problematic scale that might crash with `...` - let's change this to `Object.values(contextElements).concat(Object.values(...).concat(Object.values(...))` to be on the safe side?

**Review by netama — APPROVED** [review]
looks great. added a few suggestions but nothing major so approving, lmk if you want another look after the changes

File: packages/adapter-components/src/elements/element_getter.ts (hunk 2 above)
Comment (edenhassid) [reply packages/adapter-components/src/elements/element_getter.ts:78]: but we have the `persistInstances` field as well, how would that work?

Comment (netama) [reply packages/adapter-components/src/elements/element_getter.ts:88]: sorry missed it, so it can be `FetchElements & { persistInstances: boolean }`

File: packages/adapter-components/src/elements/swagger/fetch_type.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+/*
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/fetch_type.ts:1]: this file is empty now :) let's remove it?

File: packages/adapter-components/src/elements/swagger/instance_elements.ts [hunk 6]
```diff
@@ -488,13 +490,30 @@ export const getAllInstances = async ({
     getElemIdFunc,
   }
 
-  return getElementsWithContext({
+  const { elements, errors } = await getElementsWithContext<InstanceElement>({
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/instance_elements.ts:493]: this can theoretically also contain `configChanges`, no? why not just `return getElementsWithContext...` as we do today?

File: packages/zendesk-adapter/src/adapter.ts [hunk 7]
```diff
@@ -28,6 +28,7 @@ import {
 import { logDuration, resolveChangeElement, resolveValues, restoreChangeElement } from '@salto-io/adapter-utils'
 import { collections, objects } from '@salto-io/lowerdash'
 import { logger } from '@salto-io/logging'
+// import { FetchElements } from '@salto-io/adapter-components/dist/src/elements/element_getter'
```
Comment (netama) [inline packages/zendesk-adapter/src/adapter.ts:31]: remove?

**Review by netama — APPROVED** [review]
added a few nits but looks great
